### PR TITLE
misc bugfixes

### PR DESCRIPTION
**Pull request by ichung**
A couple of small changes I made in the process of building and deploying a cluster.  Mostly bugfixes, some QoL improvements.

---

Checklist:

 - [X] I have split up this change into one or more appropriately-delineated commits.
 - [X] The first line of each commit is of the form "[component]: do something"
 - [X] I have written a complete, multi-line commit message for each commit.
 - [X] I have formatted any Go code that I have changed with gofmt.
 - [x] I have signed each commit with my GPG key.
 - [x] My changes have passed CI.
 - [x] I have built my changes locally, and tested that the changes work as expected.
 - [x] I have deployed a cluster incorporating my changes, and checked that it deploys successfully.
 - [x] I have assigned this issue to an appropriate reviewer. (Choose @celskeggs if you are not otherwise certain.)
 - [X] I consider my PR complete and ready to be merged without my further input, assuming that it passes CI and code review.

### CODE REVIEW

File: create-chroot.sh [hunk 1]
```diff
@@ -29,12 +29,13 @@ then
...
 fi
 
-mkdir "${HOMEWORLD_CHROOT}"
+mkdir -m 'u=rwx,go=rx' "${HOMEWORLD_CHROOT}"
 sudo debootstrap --include="$(cat chroot-packages.list | grep -vE '^#' | tr '\n ' ',,' | sed 's/,$//' | sed 's/,,/,/g')" stretch "${HOMEWORLD_CHROOT}" http://debian.csail.mit.edu/debian/
 sudo ln -sT "homeworld/building" "${HOMEWORLD_CHROOT}/h"
 sudo ln -sT /h/glass/glass.py "${HOMEWORLD_CHROOT}/usr/local/bin/glass"
 sudo chroot "${HOMEWORLD_CHROOT}" groupadd "$(id -gn)" -g "$(id -g)"
 sudo chroot "${HOMEWORLD_CHROOT}" useradd -m -u "$(id -u)" -g "$(id -g)" "$USER" -s "/bin/bash"
+sudo chroot "${HOMEWORLD_CHROOT}" pip install wheel
```
Comment (celskeggs) [inline create-chroot.sh]: Is it not enough to add the debian package for wheel?

File: create-chroot.sh [hunk 2]
```diff
@@ -11,7 +11,7 @@ then
     exit 1
 fi
 
-if [ "${HOMEWORLD_CHROOT}" =~ " " ]
+if [[ "${HOMEWORLD_CHROOT}" =~ " " ]]
 then
     echo "chroot name canot include a space" 1>&2
```
Comment (celskeggs) [inline create-chroot.sh]: If you want to, you could also include a fix for this typo.

**Comment by celskeggs** [comment]
Poke me once you've completed the checklist.

File: building/glass/upload.py [hunk 3]
```diff
@@ -123,6 +123,18 @@ def gs_rsync(local_path: str, remote_path: str, boto_path: str):
...
 
+gs_branch_regex = re.compile(r'([a-z0-9_.-]+)((/[a-z0-9/_.-]*)?)')
+
+def gs_upload_path(branch):
+    m = gs_branch_regex.fullmatch(branch)
+    if m is None:
+        raise ValueError('invalid branch name {}'.format(branch))
+    domain, path = m.group(1), m.group(2)
+    if domain.endswith('.storage.googleapis.com'):
+        domain = domain[:-len('.storage.googleapis.com')]
```
Comment (celskeggs) [inline building/glass/upload.py]: Any reason not to just use basic manipulation options?

i.e. `domain, path = branch.split('/', 1)` and then `if domain.endswith('.storage.googleapis.com'): domain = domain.rsplit('/', 3)[0]`

Comment (ichung) [reply building/glass/upload.py]: the regex also validates the form of the URL

File: building/glass/upload.py [hunk 4]
```diff
@@ -98,15 +98,13 @@ def gs_rsync(local_path: str, remote_path: str, boto_path: str):
     # TODO: do this directly rather than by shelling out
     env = dict(os.environ)
     env["BOTO_PATH"] = boto_path
-    subprocess.check_call(["gsutil", "-h", "Cache-Control:private, max-age=0, no-transform", "-m", "rsync", "-d", "-r", "-c", local_path, "gs://" + remote_path], env=env)
+    subprocess.check_call(["gsutil", "-h", "Cache-Control:private, max-age=0, no-transform", "-m", "rsync", "-d", "-r", "-c", local_path, remote_path], env=env)
 
 
-GS_BRANCH_REGEX = re.compile("[a-z0-9-]+[.][a-z0-9.-]+/[a-z0-9-]+")
-
+GS_BRANCH_REGEX = re.compile(r'([a-z0-9_.-]+)((/[a-z0-9/_.-]*)?)')
```
Comment (celskeggs) [inline building/glass/upload.py]: Can you use a normal string literal instead of a raw string literal?

File: building/glass/aptbranch.py [hunk 5]
```diff
@@ -63,6 +63,13 @@ def select_upload_config(upload_targets: list, name: str):
 
 
 class Config:
+    @classmethod
+    def list_branches(cls):
+        if not os.path.exists(CONFIG_PATH):
+            raise Exception("cannot find branches config at %s, use %s.example to create one" % (CONFIG_PATH, CONFIG_PATH))
+        branches_config = validate.load_validated(CONFIG_PATH, CONFIG_SCHEMA_NAME)["branches"]
+        return [config['name'] for config in branches_config]
+
```
Comment (celskeggs) [inline building/glass/aptbranch.py]: I wonder if there's a way to not duplicate code between here and `__init__`?

File: building/components/homeworld-admin-tools/src/infra.py [hunk 6]
```diff
@@ -22,16 +22,15 @@ def infra_admit(server_principal: str) -> None:
...
         token = admit(node.hostname)
-        tokens[node.hostname] = (node.kind, node.ip, token)
-    print("host".center(16, "="), "kind".center(8, "="), "ip".center(14, "="), "token".center(23, "="))
-    for key, (kind, ip, token) in sorted(tokens.items()):
-        print(key.rjust(16), kind.center(8), str(ip).center(14), token.ljust(23))
-    print("host".center(16, "="), "kind".center(8, "="), "ip".center(14, "="), "token".center(23, "="))
+        tokens.append((node.hostname, node.kind, node.ip, token))
+    print('{:=^16}{:=^8}{:=^14}{:=^23}'.format('host', 'kind', 'ip', 'token'))
+    for hostname, kind, ip, token in tokens:
+        print('{:>16}{:^8}{:^14}{:<23}'.format(hostname, kind, str(ip), token))
```
Comment (celskeggs) [inline building/components/homeworld-admin-tools/src/infra.py]: Doesn't this change remove the spacing between columns?

    >>> print("host".center(16, "="), "kind".center(8, "="), "ip".center(14, "="), "token".center(23, "="))
    ======host====== ==kind== ======ip====== =========token=========
    >>> print('{:=^16}{:=^8}{:=^14}{:=^23}'.format('host', 'kind', 'ip', 'token')) 
    ======host========kind========ip===============token=========

File: building/components/homeworld-admin-tools/src/infra.py [hunk 7]
```diff
@@ -22,16 +22,15 @@ def infra_admit(server_principal: str) -> None:
...
         token = admit(node.hostname)
-        tokens[node.hostname] = (node.kind, node.ip, token)
-    print("host".center(16, "="), "kind".center(8, "="), "ip".center(14, "="), "token".center(23, "="))
-    for key, (kind, ip, token) in sorted(tokens.items()):
-        print(key.rjust(16), kind.center(8), str(ip).center(14), token.ljust(23))
-    print("host".center(16, "="), "kind".center(8, "="), "ip".center(14, "="), "token".center(23, "="))
+        tokens.append((node.hostname, node.kind, node.ip, token))
+    print('{:=^16} {:=^8} {:=^14} {:=^23}'.format('host', 'kind', 'ip', 'token'))
+    for hostname, kind, ip, token in tokens:
+        print('{:>16} {:^8} {:^14} {:<23}'.format(hostname, kind, str(ip), token))
```
Comment (celskeggs) [inline building/components/homeworld-admin-tools/src/infra.py]: Is removing the table footer an intentional change?

File: building/glass/aptbranch.py [hunk 8]
```diff
@@ -64,16 +64,17 @@ def select_upload_config(upload_targets: list, name: str):
...
+    @classmethod
+    def list_branches(cls):
+        return [config['name'] for config in cls.load_configs()['branches']]
 
     def __init__(self, branch: str):
-        if not os.path.exists(CONFIG_PATH):
-            raise Exception("cannot find branches config at %s, use %s.example to create one" % (CONFIG_PATH, CONFIG_PATH))
-        configs = validate.load_validated(CONFIG_PATH, CONFIG_SCHEMA_NAME)
+        configs = type(self).load_configs()
```
Comment (celskeggs) [inline building/glass/aptbranch.py]: Don't indirect through type(self); just call `Config.load_configs()`.

**Pull request merged**